vhost-device-spi/src/vhu_spi.rs

```diff
@@ -0,0 +1,948 @@
+
+            (tx_buf, rx_buf, trans_len) = match descriptors.len() {
+                3 => {
+                    // half-duplex transfer
+                    let desc_buf = descriptors[1];
+                    let len = desc_buf.len();
+
+                    if len == 0 {
+                        return Err(Error::UnexpectedDescriptorSize(1, len));
+                    }
+
+                    if desc_buf.is_write_only() {
+                        rx_buf = vec![0; len as usize];
+                        tx_buf = Vec::<u8>::new();
+                    } else {
+                        tx_buf = vec![0; len as usize];
+
+                        desc_chain
+                            .memory()
+                            .read(&mut tx_buf, desc_buf.addr())
+                            .map_err(|_| Error::DescriptorReadFailed)?;
+
+                        rx_buf = Vec::<u8>::new();
+                    }
+
+                    (tx_buf, rx_buf, len)
+                }
+
+                4 => {
+                    // full-duplex transfer
+                    let tx_desc_buf = descriptors[1];
+                    let rx_desc_buf = descriptors[2];
+
+                    let tx_len = tx_desc_buf.len();
+                    let rx_len = rx_desc_buf.len();
+
+                    if tx_len == 0 {
+                        return Err(Error::UnexpectedDescriptorSize(1, tx_len));
+                    }
+
+                    if rx_len == 0 {
+                        return Err(Error::UnexpectedDescriptorSize(1, rx_len));
+                    }
+
+                    if tx_len != rx_len {
+                        return Err(Error::TxRxTrnasLenNotEqual(tx_len, rx_len));
+                    }
+
+                    rx_buf = vec![0; rx_len as usize];
+                    tx_buf = vec![0; tx_len as usize];
+
+                    desc_chain
+                        .memory()
+                        .read(&mut tx_buf, tx_desc_buf.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+
+                    (tx_buf, rx_buf, tx_len)
+                }
+
+                _ => {
+                    return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+                }
+            };
+
+            reqs.push(SpiTransReq {
+                tx_buf,
+                rx_buf,
+                trans_len,
+                speed_hz: out_hdr.freq.to_native(),
+                delay_usecs: (out_hdr.cs_delay_hold_ns.to_native() / 1000) as u16,
+                bits_per_word: out_hdr.bits_per_word,
+                cs_change: out_hdr.cs_change,
+                tx_nbits: out_hdr.tx_nbits,
+                rx_nbits: out_hdr.rx_nbits,
+                word_delay_usecs: (out_hdr.word_delay_ns.to_native() / 1000) as u8,
+                mode: out_hdr.mode.to_native(),
+                cs_id: out_hdr.chip_select_id,
+            });
+        }
+
+        let in_hdr = match self.spi_ctrl.check_trans_params(&mut reqs[0]) {
```

Comment:
I am probably missing something here but why you are only checking only `reqs[0]`? If I understand correctly, `reqs[]` is a queue that contains all the requests that have been pushed in the the previous step, am I right?

Reply:
Exactly.
I misunderstood something here, I will update to check all requests' parameters.

Reply:
Resolved.